Blockchains > Get Started > Introduction to Polkadot SDK (#155)
polkadot-developers/polkadot-docs: develop/blockchains/get-started/intro-polkadot-sdk.md

- Networking and peer-to-peer communication (powered by [Libp2p](/polkadot-protocol/glossary#libp2p){target=\_blank})
- Consensus protocols, such as [BABE](/polkadot-protocol/glossary#blind-assignment-of-blockchain-extension-babe){target=\_blank}, [GRANDPA](/polkadot-protocol/glossary#grandpa){target=\_blank}, or [Aura](/polkadot-protocol/glossary#authority-round-aura){target=\_blank}
- Cryptography
- The ability to create portable Wasm runtimes
- A selection of pre-built modules, called [pallets](/polkadot-protocol/glossary#pallet){target=\_blank}
- Benchmarking and testing suites

!!!note
    For an in-depth dive into the monorepo, the [Polkadot SDK Rust documentation](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/index.html){target=\_blank} is highly recommended.

## Polkadot SDK Overview

The Polkadot SDK is composed of five major components:

![](/images/develop/blockchains/get-started/intro-polkadot-sdk/polkadot-sdk-structure.webp)

> - [Substrate](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/substrate/index.html){target=\_blank} - a set of libraries and primitives for building blockchains
> - [FRAME](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/frame_runtime/index.html){target=\_blank} - a blockchain development framework built on top of Substrate
> - [Cumulus](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/cumulus/index.html){target=\_blank} - a set of libraries and pallets to add parachain capabilities to a Substrate/FRAME runtime
> - [XCM (Cross Consensus Messaging)](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/xcm/index.html){target=\_blank} - the primary format for conveying messages between parachains 
> - [Polkadot](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/polkadot/index.html){target=\_blank} - the node implementation for the Polkadot protocol

Reviewer: ```suggestion
- [**Substrate**](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/substrate/index.html){target=\_blank} - a set of libraries and primitives for building blockchains
- [**FRAME**](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/frame_runtime/index.html){target=\_blank} - a blockchain development framework built on top of Substrate
- [**Cumulus**](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/cumulus/index.html){target=\_blank} - a set of libraries and pallets to add parachain capabilities to a Substrate/FRAME runtime
- [**XCM (Cross Consensus Messaging)**](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/xcm/index.html){target=\_blank} - the primary format for conveying messages between parachains 
- [**Polkadot**](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/polkadot/index.html){target=\_blank} - the node implementation for the Polkadot protocol
```

### Cumulus

Cumulus provides utilities and libraries to turn FRAME-based runtimes into runtimes that can be a parachain on Polkadot. Cumulus runtimes are still FRAME runtimes but contain the necessary functionality that allows that runtime to become a parachain on a relay chain.

## Why Use Polkadot SDK?

Using the Polkadot SDK, you can build application-specific blockchains without the complexity of building a blockchain from scratch or the limitations of building on a general-purpose blockchain. You can focus on crafting the business logic that makes your chain unique and innovative with the additional benefits of flexibility, upgradeability, open-source licensing, and cross-consensus interoperability.

## Create a Custom Blockchain Using the SDK

Before starting your blockchain development journey, you'll need to decide whether you want to build a standalone chain or a parachain that connects to the Polkadot network. Each path has its considerations and requirements. Once you've made this decision, follow these development stages:

```mermaid
graph LR
    A[Install the Polkadot SDK] --> B[Build the Chain]
    B --> C[Deploy the Chain]
```

> 1. [Install the Polkadot SDK](./install-polkadot-sdk.md) - set up your development environment with all necessary dependencies and tools

Reviewer: ```suggestion
1. [**Install the Polkadot SDK**](/develop/blockchains/get-started/install-polkadot-sdk/){target=\_blank} - set up your development environment with all necessary dependencies and tools
```

> 2. [Build the Chain](TODO: update-path) - learn how to create and customize your blockchain's runtime, configure pallets, and implement your chain's unique features

Reviewer: ```suggestion
2. [**Build the chain**](/develop/blockchains/get-started/build-custom-blockchains/){target=\_blank} - learn how to create and customize your blockchain's runtime, configure pallets, and implement your chain's unique features
```

> 3. [Deploy the Chain](./deploy-blockchain-to-polkadot.md) - follow the steps to launch your blockchain, whether as a standalone network or as a parachain on Polkadot

Reviewer: ```suggestion
3. [Deploy the chain](/develop/blockchains/get-started/deploy-blockchain-to-polkadot/){target=\_blank} - follow the steps to launch your blockchain, whether as a standalone network or as a parachain on Polkadot
```

---
title: Introduction to Polkadot SDK
description: Learn about the Polkadot SDK, what it is comprised of, and how it powers and forms the foundation of the Polkadot protocol
---

# Introduction to Polkadot SDK

## Introduction

> The [Polkadot SDK](https://github.com/paritytech/polkadot-sdk){target=\_blank} is a powerful and versatile developer kit designed to facilitate building on the Polkadot network. It provides the necessary components for creating custom blockchains, parachains, generalized rollups, and more. Written in the Rust Programming Language, it puts security and robustness at the forefront of its design.

Reviewer 2: I'm wondering about the usage of the "rollups" word here...

> Whether you're building a standalone chain or deploying a parachain on Polkadot, this SDK equips developers with the libraries and tools needed to manage runtime logic, compile the codebase, and utilize core features like staking, governance, and Cross-Consensus Messaging (XCM). It also provides a means for building generalized peer-to-peer systems, whether a blockchain or an alternative technology. The Polkadot SDK houses the following overall functionality:

Reviewer 2: > It also provides a means for building generalized peer-to-peer systems, whether a blockchain or an alternative technology

This is a bit weird for me... what would be an example of an "alternative technology"? maybe we can get rid of this sentence directly, wdyt?

> Whether you're building a standalone chain or deploying a parachain on Polkadot, this SDK equips developers with the libraries and tools needed to manage runtime logic, compile the codebase, and utilize core features like staking, governance, and Cross-Consensus Messaging (XCM). It also provides a means for building generalized peer-to-peer systems, whether a blockchain or an alternative technology. The Polkadot SDK houses the following overall functionality:

Author: > > It also provides a means for building generalized peer-to-peer systems, whether a blockchain or an alternative technology
> 
> This is a bit weird for me... what would be an example of an "alternative technology"? maybe we can get rid of this sentence directly, wdyt?

From my understanding, using the core libraries provided by the SDK, you could theoretically create a p2p system that is not necessarily a blockchain

> - Networking and peer-to-peer communication (powered by [Libp2p](/polkadot-protocol/glossary#libp2p))
> - Consensus protocols, such as [BABE](/polkadot-protocol/glossary#blind-assignment-of-blockchain-extension-babe), [GRANDPA](/polkadot-protocol/glossary#grandpa), or [Aura](/polkadot-protocol/glossary#authority-round-aura)
> - Cryptography
> - A selection of pre-built modules, called [pallets](/polkadot-protocol/glossary#pallet)
> - Benchmarking and testing suites
> - The ability to create portable Wasm runtimes

Reviewer 2: Just being nit-picky here, but for the sake of some sort of "semantic order", maybe we can list these items in the following way:

```suggestion
- Networking and peer-to-peer communication (powered by [Libp2p](/polkadot-protocol/glossary#libp2p){target=\_blank})
- Consensus protocols, such as [BABE](/polkadot-protocol/glossary#blind-assignment-of-blockchain-extension-babe){target=\_blank}, [GRANDPA](/polkadot-protocol/glossary#grandpa){target=\_blank}, or [Aura](/polkadot-protocol/glossary#authority-round-aura){target=\_blank}
- Cryptography
- The ability to create portable Wasm runtimes
- A selection of pre-built modules, called [pallets](/polkadot-protocol/glossary#pallet){target=\_blank}
- Benchmarking and testing suites
```

Reviewer 2: Also, I'm trying to think of a synonym for "Wasm runtimes" here...

maybe "Create portable custom runtimes using Wasm"?

!!!note
    For an in-depth dive into the monorepo, the [Polkadot SDK Rust documentation](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/index.html){target=\_blank} is highly recommended.

## Polkadot SDK Overview

The Polkadot SDK is composed of five major components:

> ![](/images/develop/blockchains/get-started/intro-polkadot-sdk/polkadot-sdk-structure.webp)

Reviewer 2: If we replace this image with a diagram created by us, I'd highly recommend encapsulating the five components into the Polkadot SDK.

> - [Substrate](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/substrate/index.html){target=\_blank} - a set of libraries for building blockchains

Reviewer 2: ```suggestion
- [Substrate](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/substrate/index.html){target=\_blank} - a set of libraries and primitives for building blockchains
```

- [Polkadot](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/polkadot/index.html){target=\_blank} - the node implementation for the Polkadot protocol

### Substrate

Substrate is a Software Development Kit (SDK) that uses Rust-based libraries and tools to enable you to build application-specific blockchains from modular and extensible components. Application-specific blockchains built with Substrate can run as standalone services or in parallel with other chains to take advantage of the shared security provided by the Polkadot ecosystem. Substrate includes default implementations of the core components of the blockchain infrastructure to allow you to focus on the application logic.

Every blockchain platform relies on a decentralized network of computers—called nodes—that communicate with each other about transactions and blocks.
In general, a node in this context is the software running on the connected devices rather than the physical or virtual machine in the network.

> As software, Substrate-based nodes consist of two main parts with separate responsibilities:

Author: We should create a diagram similar to the following for this section:
<img width="751" alt="image" src="https://github.com/user-attachments/assets/9553d519-afc9-4981-b54e-b8c1c6ccd7c6">

Reviewer 3: I don't like how this one looks at all.  If you want to get this on a list of diagrams for after launch, please open a ticket to request a mermaid diagram, paste the text including the information that goes in the diagram (what are we trying to explain visually?), let me know which page it's for, and assign to me. If we find we can make some before launch then we will but I don't think these should keep us from shipping.

- **Client** - services to handle network and blockchain infrastructure activity
    - Native Binary
    - Executes the Wasm runtime

>     - Everything else: Database, Networking, Mempool, Consensus, and more

Reviewer 2: ```suggestion
    - Manages components like Database, Networking, Mempool, Consensus, and others
```

>     - Native Binary

Reviewer 4: ```suggestion
    - Native binary
```

    - Also known as Host
- **Runtime** - business logic for state transitions
    - Application logic

>     - Wasm

Reviewer 2: ```suggestion
    - Compiled to [Wasm](https://webassembly.org/){target=\_blank}
```

>     - Manages components like Database, Networking, Mempool, Consensus, and others

Reviewer 4: Don't need to capitalize terms like mempool, consensus, unless it is a name of specific type of consensus, .e.g Babe or Grandpa, etc.

    - Stored as a part of the chain state
    - Also known as STF (State Transition Function)

### FRAME

FRAME provides the core modular and extensible components that make the Substrate SDK flexible and adaptable to different use cases.
FRAME includes Rust-based libraries that simplify the development of application-specific logic.

> Most of the functionality that FRAME provides takes the form of plug-in modules called [pallets](/polkadot-protocol/glossary#pallet) that you can add and configure to suit your requirements.

Reviewer 2: ```suggestion
Most of the functionality that FRAME provides takes the form of plug-in modules called [pallets](/polkadot-protocol/glossary#pallet){target=\_blank} that you can add and configure to suit your requirements.
```

> 

Author: Here we could also use a diagram like this:
<img width="936" alt="image" src="https://github.com/user-attachments/assets/4bdbf23f-7b01-41af-9146-c72e2ee18855">
Or
<img width="605" alt="image" src="https://github.com/user-attachments/assets/0a0d4def-3801-4bfd-8a7c-efcf71a5d4fc">

Reviewer 3: This one is nicer. Do you think we need this before launch or can it go on the list for after?

Author: We can add it later. I've created tickets in Jira

### Cumulus

Cumulus provides utilities and libraries to turn FRAME-based runtimes into runtimes that can be a parachain on Polkadot. Cumulus runtimes are still FRAME runtimes but contain the necessary functionality that allows for that runtime to become a parachain on a relay chain.

## Why Use Polkadot SDK?

Using the Polkadot SDK, you can build application-specific blockchains without the complexity of building a blockchain from scratch or the limitations of building on a general-purpose blockchain.
You can focus on crafting the business logic that makes your chain unique and innovative with the additional benefits of flexibility, upgradeability, open-source licensing, and cross-consensus interoperability.

## How to Create a Custom Blockchain using the SDK?

Before starting your blockchain development journey, you'll need to decide whether you want to build a standalone chain or a parachain that connects to the Polkadot network. Each path has its considerations and requirements. Once you've made this decision, follow these development stages:

```mermaid
graph LR
    A[Install] --> B[Build]

>     B --> C[Deploy]

Reviewer 2: Would it make sense to make the diagram entities' names match with the title below?
like:

```mermaid
graph LR
    A[Install the Polkadot SDK] --> B[Build the Chain]
    B --> C[Deploy the Chain]
```

Author: Yes it does, I've updated it

```

> 1. [Install the Polkadot SDK](./install-polkadot-sdk.md) - set up your development environment with all necessary dependencies and tools
> 
> 2. [Build the Chain](TODO: update-path) - learn how to create and customize your blockchain's runtime, configure pallets, and implement your chain's unique features
> 
> 3. [Deploy the Chain](./deploy-blockchain-to-polkadot.md) - follow the steps to launch your blockchain, whether as a standalone network or as a parachain on Polkadot
> 

Reviewer 2: ```suggestion
1. [Install the Polkadot SDK](./install-polkadot-sdk.md){taget=\_blank} - set up your development environment with all necessary dependencies and tools

2. [Build the Chain](TODO: update-path){target=\_blank} - learn how to create and customize your blockchain's runtime, configure pallets, and implement your chain's unique features

3. [Deploy the Chain](./deploy-blockchain-to-polkadot.md){target=\_blank} - follow the steps to launch your blockchain, whether as a standalone network or as a parachain on Polkadot

```

Author: I don't think we actually want to use `target=\_blank` here because we want users to go to the other page
